Mathlib/Data/Real/IsNonarchimedean.lean

```diff
@@ -0,0 +1,152 @@
+`f (r + s) ≤ max (f r) (f s)` for all `r s : R`. This file proves basic properties of
+nonarchimedean functions.
+-/
+
+/-- A function `f : R → ℝ≥0` is nonarchimedean if it satisfies the strong triangle inequality
+  `f (r + s) ≤ max (f r) (f s)` for all `r s : R`. -/
+def IsNonarchimedean {A : Type*} [Add A] (f : A → ℝ) : Prop :=
+  ∀ r s, f (r + s) ≤ max (f r) (f s)
+
+namespace IsNonarchimedean
+
+/-- A nonarchimedean function satisfies the triangle inequality. -/
+theorem add_le {α : Type*} [AddCommGroup α] {f : α → ℝ} (hf : ∀ x : α, 0 ≤ f x)
+    (hna : IsNonarchimedean f) (a b : α) : f (a + b) ≤ f a + f b := by
+  apply le_trans (hna _ _)
+  rw [max_le_iff, le_add_iff_nonneg_right, le_add_iff_nonneg_left]
+  exact ⟨hf _, hf _⟩
+
+/-- If `f` is a nonarchimedean additive group seminorm on `α`, then for every `n : ℕ` and `a : α`,
+  we have `f (n • a) ≤ (f a)`. -/
+theorem nsmul_le {F α : Type*} [AddCommGroup α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] {f : F} (hna : IsNonarchimedean f) (n : ℕ) (a : α) :
+    f (n • a) ≤ f a := by
+  induction n with
+  | zero => rw [zero_nsmul, map_zero _]; exact apply_nonneg _ _
+  | succ n hn =>
+    rw [add_smul, one_smul]
+    exact le_trans (hna _ _) (max_le hn (le_refl _))
+
+/-- If `f` is a nonarchimedean additive group seminorm on `α`, then for every `n : ℕ` and `a : α`,
+  we have `f (n * a) ≤ (f a)`. -/
+theorem nmul_le {F α : Type*} [Ring α] [FunLike F α ℝ] [AddGroupSeminormClass F α ℝ]
+    {f : F} (hna : IsNonarchimedean f) (n : ℕ) (a : α) : f (n * a) ≤ f a := by
+  rw [← nsmul_eq_mul]
+  exact nsmul_le hna _ _
+
+/-- If `f` is a nonarchimedean additive group seminorm on `α` and `x y : α` are such that
+  `f y ≠ f x`, then `f (x + y) = max (f x) (f y)`. -/
+theorem add_eq_max_of_ne {F α : Type*} [Ring α] [FunLike F α ℝ]
+    [AddGroupSeminormClass F α ℝ] {f : F} (hna : IsNonarchimedean f) {x y : α} (hne : f y ≠ f x) :
+    f (x + y) = max (f x) (f y) := by
+  wlog hle : f y ≤ f x generalizing y x with H
+  · rw [add_comm, max_comm]
+    exact H hne.symm (le_of_not_le hle)
+  · have hlt : f y < f x := lt_of_le_of_ne hle hne
+    have : f x ≤ max (f (x + y)) (f y) :=
+      calc f x = f (x + y + -y) := by rw [add_neg_cancel_right]
+        _ ≤ max (f (x + y)) (f (-y)) := (hna _ _)
+        _ = max (f (x + y)) (f y) := by rw [map_neg_eq_map f y]
+    have hnge : f y ≤ f (x + y) := by
+      apply le_of_not_gt
+      intro hgt
+      rw [max_eq_right_of_lt hgt] at this
+      exact not_lt_of_ge this hlt
+    have : f x ≤ f (x + y) := by rwa [max_eq_left hnge] at this
+    exact le_antisymm (hna _ _) (by rwa [max_eq_left_of_lt hlt])
+
+/-- Given a nonarchimedean additive group seminorm `f` on `α`, a function `g : β → α` and a finset
+  `t : Finset β`, we can always find `b : β`, belonging to `t` if `t` is nonempty, such that
+  `f (t.sum g) ≤ f (g b)` . -/
+theorem finset_image_add {F α : Type*} [Ring α] [FunLike F α ℝ]
```

Comment:
Perhaps do you want to extract a primed lemma (or something) that assumes that `t` is nonempty, simplifying the result?

Reply:
I have added that lemma.